### ISSUE TITLE
Feature request - incremental loading

### ISSUE THREAD

**Issue by avalanche1** [meteor/meteor#6967]
Is incremental loading on agenda?
incremental loading = loading only the js\css necessary for current app state (eg route); like not bundling everything toghether for each visitor: no need for admin templates for simple visitor, no need for all templates for a landing page visitor...

**Comment by benjamn** [comment]
This is not currently on our short-term agenda, but the way we're almost certainly going to implement incremental loading is to allow `import` statements in `async` functions:
```js
import { a } from "./a";

async function later() {
  import { b } from "./b";
  return a + b;
}

later().then(result => {
  console.log(result);
});
```
Because the `async` function can wait for a different bundle of code to be loaded, modules `./a` and `./b` can be split between an early bundle and a later bundle. Or not! They could equally well remain in the same monolithic bundle, and the logic would be the same, but the app would be slower to load, especially if `later` isn't called until later.

It's important to use statically analyzable `import` statements (or `require` calls with string literal identifiers) so that the dependency scanner is able to anticipate which bundles of modules will be imported, and potentially begin downloading them as soon as the page is idle, so that they are already available by the time `later` is called. Whether this is a good idea depends on when (if ever) the `async` function is likely to be called, of course. There are lots of variables in play here, but `import` statements in `async` functions gives us all the flexibility we need to optimize for every case.

Route callbacks are a great use case for `async` functions, since it's ok if there's a small delay before the page responds to the route change.

**Comment by laosb** [comment]
I think it's not a feature in `Project:Tool`?

**Comment by gsabran** [comment]
Cool, that sounds quite similar to what is done with things like webpack that offer code splitting. Do you have thoughts on how you see something like this working:
```javascript
// router.jsx
import { main } from "./main";
import { core } from "./lib/core"

async function later() {
  import { OneView } from "./b";
}

// b
import { core } from "./lib/core"
import { bSpecific } from "./lib/bSpecific"

export const OneView = ...
```

Would things be prebundled for fast loading? If so would `core` be loaded twice? Or would there be different bundles to load according to what is needed?

**Comment by benjamn** [comment]
Assuming there are two bundles, `./lib/core` definitely has to go in the earlier bundle, and since the `later` function could assume the earlier bundle has already loaded, the later bundle would not need to include `./lib/core` again. In general, `async` functions that contain `import` statements will only need to load modules that weren't already depended upon by an outer scope.

There could be devils in the details that prevent this from working, but that's my intuition at least :)

**Comment by gsabran** [comment]
Sounds good, really looking forward to it! Do you know if it's going to be part of the 1.3/1.4 releases?

**Comment by gsabran** [comment]
Yes this is the main reason I moved all my project to webpack. I'm very surprised that this doesn't come by default with the magic Meteor's boilerplate. I'm in the process of better documenting a playground project, but if you want to have a look at webpack + code splitting + other stuff [here](https://github.com/gsabran/kickstart-meteor-react-flowrouter-SSR-codeSplit) it is:

**Comment by mitar** [comment]
Also, probably all <script> tags should load with `async` attribute.

**Comment by hwillson** [comment]
Implemented in Meteor 1.5 🎉. Closing - thanks!

**Comment by devagrawal09** [comment]
The implementation of dynamic `import()` only works for the modules written by us. Atmosphere packages are still there in the initial bundle making the first load heavy. Packages like `aldeed:autoform` that are only required in a few places are loaded with the initial bundle unnecessarily.
Is there any way atmosphere packages can be dynamically imported too? Or do we have to wait for the team to add such functionality to the build system?
